### PR TITLE
Explainer API for Local Classifier per parent node #minor

### PR DESCRIPTION
**Pull request by ashishpatel16**
Hi @mirand863, this PR aims to present the first iteration of the LCPPN explainer class that enables the calculation of Shapley values for explaining predictions.

### CODE REVIEW

**Comment by ashishpatel16** [comment]
[Here](https://colab.research.google.com/drive/1wqSl1t_Qn2f62WNZQ48mdB0mNeu1XSF1?usp=sharing) you can find a Colab notebook to test the api using a dummy hierarchical dataset!

**Comment by mirand863** [comment]
> ## [Codecov](https://app.codecov.io/gh/scikit-learn-contrib/hiclass/pull/106?src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=scikit-learn-contrib) Report
> 
> Attention: `66 lines` in your changes are missing coverage. Please review.
> 
> > Comparison is base [(`c7979b2`)](https://app.codecov.io/gh/scikit-learn-contrib/hiclass/commit/c7979b20ee897244db7933570f422b653183884a?el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=scikit-learn-contrib) 96.86% compared to head [(`9658c4a`)](https://app.codecov.io/gh/scikit-learn-contrib/hiclass/pull/106?src=pr&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=scikit-learn-contrib) 91.58%.
> 
> [Files](https://app.codecov.io/gh/scikit-learn-contrib/hiclass/pull/106?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=scikit-learn-contrib) 	Patch % 	Lines
> [hiclass/Explainer.py](https://app.codecov.io/gh/scikit-learn-contrib/hiclass/pull/106?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=scikit-learn-contrib#diff-aGljbGFzcy9FeHBsYWluZXIucHk=) 	21.68% 	[65 Missing ⚠️ ](https://app.codecov.io/gh/scikit-learn-contrib/hiclass/pull/106?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=scikit-learn-contrib)
> [hiclass/LocalClassifierPerParentNode.py](https://app.codecov.io/gh/scikit-learn-contrib/hiclass/pull/106?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=scikit-learn-contrib#diff-aGljbGFzcy9Mb2NhbENsYXNzaWZpZXJQZXJQYXJlbnROb2RlLnB5) 	94.11% 	[1 Missing ⚠️ ](https://app.codecov.io/gh/scikit-learn-contrib/hiclass/pull/106?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=scikit-learn-contrib)
> Additional details and impacted files
> 
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/scikit-learn-contrib/hiclass/pull/106?src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=scikit-learn-contrib). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=scikit-learn-contrib).

The test coverage is lower because shap and xarray are not being installed in the github actions. Can you please add them there? It would be nice to keep test coverage high if possible.

**Comment by mirand863** [comment]
@ashishpatel16 it would be great to have some documentation so the user can understand how the explainer works, how to use it and what is the expected output. For example, here we explain how the local classifier per node works https://hiclass.readthedocs.io/en/latest/algorithms/local_classifier_per_node.html and in the gallery of examples we show how to use the library https://hiclass.readthedocs.io/en/latest/auto_examples/plot_pipeline.html#sphx-glr-auto-examples-plot-pipeline-py. I guess in this PR you can already include some documentation about the explainer for the local classifier per parent node and an example in the gallery on how to use it, showing expected input, output and perhaps how to interpret it.

**Review by mirand863 — CHANGES_REQUESTED** [review]

File: hiclass/LocalClassifierPerParentNode.py [hunk 1]
```diff
@@ -224,3 +224,44 @@ def _fit_digraph(self, local_mode: bool = False, use_joblib: bool = False):
         self.logger_.info("Fitting local classifiers")
         nodes = self._get_parents()
         self._fit_node_classifier(nodes, local_mode, use_joblib)
+
+    def get_predict_proba(self, X):
```
[the remark this answers is not in the document]
Comment (ashishpatel16) [reply hiclass/LocalClassifierPerParentNode.py:247]: Sure, I will completely remove this method along with the corresponding test case.

[the remark this answers is not in the document]
Comment (ashishpatel16) [reply hiclass/LocalClassifierPerParentNode.py]: Resolved by [81817c3](https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/81817c3ce04594bc289ebe50fe9a9ef01a1d881c).

File: hiclass/LocalClassifierPerParentNode.py [hunk 2]
```diff
@@ -224,3 +224,44 @@ def _fit_digraph(self, local_mode: bool = False, use_joblib: bool = False):
...
+            to ``dtype=np.float32``. If a sparse matrix is provided, it will be
+            converted into a sparse ``csr_matrix``.
+        Returns
+        -------
+        prediction_probabilities : dict or list of dicts if X contains multiple samples.
+            Each dict contains the node as its key and prediction probabilities for
+            the predicted classes as value.
+        """
+        prediction_probabilities = []
+        for sample in X:
```
[the remark this answers is not in the document]
Comment (ashishpatel16) [reply hiclass/LocalClassifierPerParentNode.py]: I have completely removed this function with this [commit](https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/81817c3ce04594bc289ebe50fe9a9ef01a1d881c) since it is not needed anymore.

File: Pipfile [hunk 3]
```diff
@@ -7,6 +7,8 @@ name = "pypi"
 networkx = "*"
 numpy = "*"
 scikit-learn = "*"
+shap = "*"
+xarray = "*"
```
[the remark this answers is not in the document]
Comment (ashishpatel16) [reply Pipfile]: Resolved by [9ba6f28](https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/9ba6f28decebf93718c1c1ec5308ace0802154aa).

File: hiclass/Explainer.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+            explanation = self._calculate_shap_values(sample.reshape(1, -1))
+            explanations.append(explanation)
+        return explanations
+
+    def _get_traversed_nodes(self, samples):
+        # Helper function to return traversed nodes
+        if isinstance(self.hierarchical_model, LocalClassifierPerParentNode):
+            traversals = []
+            start_node = self.hierarchical_model.root_
+            for x in samples:
```
[the remark this answers is not in the document]
Comment (ashishpatel16) [reply hiclass/Explainer.py]: Resolved by [1b90da8](https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/1b90da80f0f02bd0638e1522abe89a707f08e3d2).

File: README.md [hunk 5]
```diff
@@ -199,6 +199,38 @@ pipeline.fit(X_train, Y_train)
...
+classifier.fit(X_train, Y_train)
+
+# Predict
+predictions = classifier.predict(X_test)
+
+# Explain
+explainer = Explainer(classifier, data=X_train, mode="tree")
+explanations = explainer.explain(X_test)
+```
+
```
Comment (mirand863) [inline README.md:233]: I don't think this belongs in the quick start in the README. Please, note that I only put a simple example here in order not to overwhelm the reader with information. All other features I added an index in the features section with a hyperlink to their respective examples in the gallery of examples. You can also put hyperlinks to the google colabs notebooks there.

File: docs/examples/plot_lcppn_explainer.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+)
+X_test = np.array([[35.5, 0.0, 1.0, 1.0, 3.0, 3.0, 0.0, 2.0, 37.5]])
+Y_train = np.array(
+    [
+        ["Gastrointestinal", "Norovirus", ""],
+        ["Respiratory", "Covid", ""],
+        ["Allergy", "External", "Bee Allergy"],
+        ["Respiratory", "Cold", ""],
+    ]
+)
```
Comment (mirand863) [inline docs/examples/plot_lcppn_explainer.py:31]: is it possible to create a function to download and load your platypus dataset? if it is not too large and can be executed quickly on read the docs, I would suggest doing something similar to this example from [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_iris.html#sklearn.datasets.load_iris):

```python
>>> from sklearn.datasets import load_iris
>>> data = load_iris()
>>> data.target[[10, 25, 50]]
array([0, 0, 1])
>>> list(data.target_names)
['setosa', 'versicolor', 'virginica']
```

Also maybe you can add some figures here like the ones you have on colabs.

File: docs/requirements.txt [hunk 7]
```diff
@@ -9,3 +9,5 @@ pandas==1.4.2
 ray==1.13.0
 numpy
 git+https://github.com/charles9n/bert-sklearn.git@master
+shap==0.44.1
+xarray
```
Comment (mirand863) [inline docs/requirements.txt:13]: might be a good idea to set the version for xarray too

File: hiclass/Explainer.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        Parameters
+        ----------
+        hierarchical_model : HierarchicalClassifier
+            The hierarchical classification model to explain.
+        data : array-like or None, default=None
+            The dataset used for creating the SHAP explainer.
+        n_jobs : int, default=1
+            The number of jobs to run in parallel.
+        algorithm : str, default="auto"
+            The algorithm to use for SHAP explainer. Possible values are 'linear', 'tree', 'auto', 'permutation'. or 'partition'
```
Comment (mirand863) [inline hiclass/Explainer.py:52]:
```suggestion
            The algorithm to use for SHAP explainer. Possible values are 'linear', 'tree', 'auto', 'permutation' or 'partition'
```

File: hiclass/Explainer.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+        >>> explanations = explainer.explain(x_test)
+        <xarray.Dataset>
+        Dimensions:          (class: 3, sample: 1, level: 2, feature: 2)
+        Coordinates:
+          * class            (class) <U1 '1' '3' '4'
+          * level            (level) int64 0 1
+        Dimensions without coordinates: sample, feature
+        Data variables:
+            node             (sample, level) <U13 'hiclass::root' '3'
+            predicted_class  (sample, level) <U24 '3' '3::HiClass::Separator::4'
```
Comment (mirand863) [inline hiclass/Explainer.py:77]:
```suggestion
            predicted_class  (sample, level) <U24 '3' '4'
```
is it possible to remove the separator from the output shown to the user? It is meant to be for internal use only and transparent to the user. Besides, I am planning to remove it in the future.

File: docs/source/get_started/install.rst [hunk 10]
```diff
@@ -7,6 +7,17 @@ To install HiClass from the Python Package Index (PyPI) simply run:
...
+Additionally, it is also possible to install optional packages along. To install optional packages run:
+
+.. code-block:: bash
+
+    pip install hiclass"[<extra_name>]"
+
+:literal:`<extra_name>` can have one of the following options:
+
+- ray: Installs the ray package, which is required for parallel processing support.
+- xai: Installs the shap and xarray packages, which are required for explaining Hiclass predictions.
```
Comment (mirand863) [inline docs/source/get_started/install.rst:19]:
```suggestion
- xai: Installs the shap and xarray packages, which are required for explaining HiClass' predictions.
```

**Review by mirand863 — CHANGES_REQUESTED** [review]

File: hiclass/Explainer.py [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+                            self.hierarchical_model.hierarchy_.successors(predecessor)
+                        )
+                        if len(successors) > 0:
+                            classifier = self.hierarchical_model.hierarchy_.nodes[
+                                predecessor
+                            ]["classifier"]
+                            traversals[mask, level] = classifier.predict(
+                                predecessor_x
+                            ).flatten()
+            return traversals
```
Comment (mirand863) [inline hiclass/Explainer.py:192]: If there are different implementations of this method for each local classifier, I would recommend to either use inheritance or call another method with the different implementations. Otherwise I am afraid this method will get way too large in the next pull requests.

File: tests/test_LocalClassifierPerParentNode.py [hunk 12]
```diff
@@ -10,7 +10,7 @@
 from sklearn.linear_model import LogisticRegression
 from sklearn.utils.estimator_checks import parametrize_with_checks
 from sklearn.utils.validation import check_is_fitted
-
+from sklearn.ensemble import RandomForestClassifier
```
Comment (mirand863) [inline tests/test_LocalClassifierPerParentNode.py:13]:
```suggestion
```
Maybe you don't need this import anymore

File: hiclass/Explainer.py (hunk 11 above)
Comment (ashishpatel16) [reply hiclass/Explainer.py]: Update: Now instead of `_get_traversed_nodes` I will write three separate method calls  `_get_traversed_nodes_lcppn`, `_get_traversed_nodes_lcpn`, and `_get_traversed_nodes_lcpl`.

Each representing the individual implementation for the corresponding hiclass model. The `_calculate_shap_values` method now calls the appropriate method by a `if-elif` construct which checks the hiclass model using `isinstance() util.

Resolved by [ada4e88](https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/ada4e88880a486099487a81cde0eb9b3ebec59ca).

File: docs/source/get_started/install.rst (hunk 10 above)
Comment (ashishpatel16) [reply docs/source/get_started/install.rst:19]: I have included the exact version in `setup.py` as well.

Resolved by [a0d3f59](https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/a0d3f59e05ce4e90eaeee2be791925b735a0f2b8).

File: docs/requirements.txt (hunk 7 above)
Comment (ashishpatel16) [reply docs/requirements.txt]: I have included the exact version in setup.py as well.

Resolved by [a0d3f59](https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/a0d3f59e05ce4e90eaeee2be791925b735a0f2b8).

File: docs/source/get_started/install.rst (hunk 10 above)
Comment (ashishpatel16) [reply docs/source/get_started/install.rst]: Resolved by [ae48bf0](https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/ae48bf02a40d4cfcac8aae4586b1107d529167e3).

File: README.md (hunk 5 above)
Comment (ashishpatel16) [reply README.md]: The README has been updated - 
- Added Hierarchical Explainability in the features subsection.
- **Removed** code example from Quick start.
- Added some text with links to colab notebook (LCPPN only with this PR) and link to Hierarchical Explainability page on Read the Docs.

Marking resolved (by https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/ae48bf02a40d4cfcac8aae4586b1107d529167e3).

File: docs/examples/plot_lcppn_explainer.py (hunk 6 above)
Comment (ashishpatel16) [reply docs/examples/plot_lcppn_explainer.py]: I have updated the `plot_lcppn_explainer.py` example to fetch a CSV file (containing ~800 entries). The example now uses the platypus diseases dataset to generate the shape values. One visualization example has also been included.

You can check the rendered example on Read the Docs [here](https://hiclass--106.org.readthedocs.build/en/106/auto_examples/plot_lcppn_explainer.html#sphx-glr-auto-examples-plot-lcppn-explainer-py).

Resolved by https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/820c1a52370d8e37d36d285ef059a62e789b1850.

File: README.md [hunk 13]
```diff
@@ -34,6 +34,7 @@ HiClass is an open-source Python library for hierarchical classification compati
 - **[Hierarchical metrics](https://hiclass.readthedocs.io/en/latest/api/utilities.html#hierarchical-metrics):** HiClass supports the computation of hierarchical precision, recall and f-score, which are more appropriate for hierarchical data than traditional metrics.
 - **[Compatible with pickle](https://hiclass.readthedocs.io/en/latest/auto_examples/plot_model_persistence.html):** Easily store trained models on disk for future use.
 - **[BERT sklearn](https://hiclass.readthedocs.io/en/latest/auto_examples/plot_bert.html):** Compatible with the library [BERT sklearn](https://github.com/charles9n/bert-sklearn).
+- **[Hierarchical Explanability]():**  HiClass allows explaining hierarchical models using the [SHAP](https://github.com/shap/shap) package.
```
Comment (mirand863) [inline README.md:37]:
```suggestion
- **[Hierarchical Explanability](TODO: add link here):**  HiClass allows explaining hierarchical models using the [SHAP](https://github.com/shap/shap) package.
```
It is still missing the hyperlink to read the docs

File: README.md [hunk 14]
```diff
@@ -199,6 +200,9 @@ pipeline.fit(X_train, Y_train)
 predictions = pipeline.predict(X_test)
 ```
 
+## Explaining Hierarchical Classifiers
```
Comment (mirand863) [inline README.md:203]: This new section is missing a hyperlink in the quick links in the beginning of the README

Comment (ashishpatel16) [reply README.md:204]: Resolved by https://github.com/scikit-learn-contrib/hiclass/pull/106/commits/af7e83abbebd90e14fbc9abf380d4c68eada4e92.

File: README.md (hunk 13 above)
Comment (ashishpatel16) [reply README.md]: Resolved by https://github.com/scikit-learn-contrib/hiclass/commit/af7e83abbebd90e14fbc9abf380d4c68eada4e92.

**Review by mirand863 — APPROVED** [review]
